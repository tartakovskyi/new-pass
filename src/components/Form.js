import React, { useEffect, useState } from 'react';
import {
  Flex,
  Button,
  FormControl,
  FormLabel,
  FormErrorMessage,
  Input,
  useToast,
} from '@chakra-ui/react';

const initialData = {
  oldPass: '',
  newPass: '',
  confirmPass: '',
};

const Form = () => {
  const [data, setData] = useState(initialData);
  const [errors, setErrors] = useState({});
  const [btnDisabled, toggleBtnDisabled] = useState(true);
  const [btnIsLoading, toggleBtnIsLoading] = useState(false);
  const toast = useToast();

  useEffect(() => {
    validate();

    if (data.newPass && data.confirmPass && data.newPass === data.confirmPass) {
      toggleBtnDisabled(false);
    } else {
      toggleBtnDisabled(true);
    }
  }, [data]);

  const handleChange = e => {
    setData({ ...data, [e.target.name]: e.target.value });
  };

  const handleSubmit = e => {
    e.preventDefault();
    const valid = validate(true);

    if (valid) {
      toggleBtnIsLoading(true);

      setTimeout(() => {
        toggleBtnIsLoading(false);

        toast({
          title: 'Password is changed',
          description: 'You can use new password for the log in',
          status: 'success',
          duration: 9000,
          isClosable: true,
        });

        setData(initialData);
      }, 1000);
    }
  };

  const validate = (submit = false) => {
    let errors = {};
    if (data.newPass) {
      if (data.newPass.length < 8) {
        errors.newPass = 'The password is too short';
      } else if (
        data.newPass.replace(/\D/g, '').length === 0 ||
        data.newPass.replace(/[^A-Za-z]/g, '').length === 0
      ) {
        errors.newPass =
          'The password must contain the both of letters and digits';
      }
    }

    if (data.newPass && data.confirmPass && data.newPass !== data.confirmPass) {
      errors.newPass = 'The password and confirm password fields do not match';
      errors.confirmPass =
        'The password and confirm password fields do not match';
    }

    if (submit) {
      Object.keys(data).forEach(field => {
        if (!data[field]) {
          errors[field] = 'The field cannot be blank';
        }
      });
    }

    setErrors(errors);

    return !Object.keys(errors).length;
  };

  return (
    <Flex
      as="form"
      w="100%"
      maxW="600px"
      direction="column"
      onSubmit={handleSubmit}
    >
      <FormControl id="oldPass" mb={5} isInvalid={errors.oldPass}>
        <FormLabel>Old password</FormLabel>
        <Input
          type="password"
          name="oldPass"
          size="lg"
          value={data.oldPass}
          onChange={handleChange}
        />
        <FormErrorMessage>{errors.oldPass}</FormErrorMessage>
      </FormControl>
      <FormControl id="newPass" mb={5} isInvalid={errors.newPass}>
        <FormLabel>New password</FormLabel>
        <Input
          type="password"
          name="newPass"
          size="lg"
          value={data.newPass}
          onChange={handleChange}
        />
        <FormErrorMessage>{errors.newPass}</FormErrorMessage>
      </FormControl>
      <FormControl id="confirmPass" mb={5} isInvalid={errors.confirmPass}>
        <FormLabel>Confirm new password</FormLabel>
        <Input
          type="password"
          name="confirmPass"
          size="lg"
          value={data.confirmPass}
          onChange={handleChange}
        />
        <FormErrorMessage>{errors.confirmPass}</FormErrorMessage>
      </FormControl>
      <Button
        type="submit"
        colorScheme="blue"
        size="lg"
        isDisabled={btnDisabled}
        isLoading={btnIsLoading}
      >
        Save changes
      </Button>
    </Flex>
  );
};

export default Form;
