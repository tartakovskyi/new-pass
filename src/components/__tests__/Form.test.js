import React from 'react';
import { render } from '@testing-library/react';
import Form from '../Form';
import {
  Flex,
  Button,
  FormControl,
  FormLabel,
  FormErrorMessage,
  Input,
  useToast,
} from '@chakra-ui/react';

const initialData = {
  oldPass: '',
  newPass: '',
  confirmPass: '',
};

test('Form should render correctly', () => {
  const { container } = render(<Form />);

  expect(container.firstChild).toMatchSnapshot();
});
