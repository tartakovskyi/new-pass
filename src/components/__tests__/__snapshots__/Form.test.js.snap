// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
<form
  class="css-kx6wjg"
>
  <div
    class="chakra-form-control css-vhwfhc"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="oldPass"
      id="oldPass-label"
    >
      Old password
    </label>
    <input
      class="chakra-input css-0"
      id="oldPass"
      name="oldPass"
      type="password"
      value=""
    />
  </div>
  <div
    class="chakra-form-control css-vhwfhc"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="newPass"
      id="newPass-label"
    >
      New password
    </label>
    <input
      class="chakra-input css-0"
      id="newPass"
      name="newPass"
      type="password"
      value=""
    />
  </div>
  <div
    class="chakra-form-control css-vhwfhc"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="confirmPass"
      id="confirmPass-label"
    >
      Confirm new password
    </label>
    <input
      class="chakra-input css-0"
      id="confirmPass"
      name="confirmPass"
      type="password"
      value=""
    />
  </div>
  <button
    class="chakra-button css-5c2176"
    disabled=""
    type="submit"
  >
    Save changes
  </button>
</form>
`;
